internal/cri/server/stats.go

```diff
@@ -0,0 +1,200 @@
+	switch m := data.(type) {
+	case *cg1.Metrics:
+		if m.CPU == nil || m.CPU.Usage == nil {
+			return fmt.Errorf("missing CPU data")
+		}
+		currentCPU = m.CPU.Usage.Total
+	case *cg2.Metrics:
+		if m.CPU == nil {
+			return fmt.Errorf("missing CPU data")
+		}
+		// convert to nano seconds
+		currentCPU = m.CPU.UsageUsec * 1000
+	case *wstats.Statistics:
+		wstats := data.(*wstats.Statistics).GetWindows()
+		if _, ok := sandboxes[stat.ID]; ok && wstats == nil {
+			// hostprocess containers sandboxes don't have stats
+			currentTimestamp = protobuf.FromTimestamp(stat.Timestamp)
+			currentCPU = 0
+		} else if wstats != nil && wstats.Processor != nil {
+			// this is slightly more accurate as it came from the windows shim
+			currentTimestamp = protobuf.FromTimestamp(wstats.Timestamp)
+			currentCPU = wstats.Processor.TotalRuntimeNS
+		} else {
+			return fmt.Errorf("invalid Windows data")
+		}
+	default:
+		return fmt.Errorf("unknown metric type")
+	}
+
+	nanoCores := calculateNanoCores(currentCPU, oldStats, currentTimestamp)
+	newStats := &stats.ContainerStats{
+		UsageCoreNanoSeconds: currentCPU,
+		UsageNanoCores:       nanoCores,
+		Timestamp:            currentTimestamp,
+	}
+	err = saveStatsFunc(stat.ID, newStats)
+	if err != nil {
+		return fmt.Errorf("failed to save container stats: %v", err)
+	}
+
+	return nil
+}
+
+func (m *metricMonitor) buildMetricsRequest() (*tasks.MetricsRequest, map[string]struct{}, map[string]struct{}) {
+	req := &tasks.MetricsRequest{}
+	sandboxes := map[string]struct{}{}
+	containers := map[string]struct{}{}
+	for _, sandbox := range m.c.sandboxStore.List() {
+		sandboxes[sandbox.ID] = struct{}{}
+		req.Filters = append(req.Filters, "id=="+sandbox.ID)
+	}
+
+	for _, cntr := range m.c.containerStore.List() {
+		containers[cntr.ID] = struct{}{}
+		req.Filters = append(req.Filters, "id=="+cntr.ID)
+	}
+	return req, sandboxes, containers
+}
+
+type saveContainerStats func(id string, containerStats *stats.ContainerStats) error
+
+func (m *metricMonitor) getExistingStats(stat *types.Metric, sandboxes map[string]struct{}) (*stats.ContainerStats, saveContainerStats, error) {
+	// it was seen as a sandbox
+	if _, ok := sandboxes[stat.ID]; ok {
+		sandbox, err := m.c.sandboxStore.Get(stat.ID)
+		if err == nil {
+			saveStatsFunc := m.c.sandboxStore.UpdateContainerStats
+			return sandbox.Stats, saveStatsFunc, err
+		}
+	}
+
+	cntr, err := m.c.containerStore.Get(stat.ID)
+	if err == nil {
+		saveStatsFunc := m.c.containerStore.UpdateContainerStats
+		return cntr.Stats, saveStatsFunc, err
+	}
+
+	// means we can't find it in either store
+	return nil, nil, fmt.Errorf("container not found: %s", stat.ID)
+}
+
+func calculateNanoCores(usageCoreNanoSeconds uint64, oldStats *stats.ContainerStats, newtimestamp time.Time) uint64 {
+	if oldStats == nil {
+		return 0
+	}
+
+	nanoSeconds := newtimestamp.UnixNano() - oldStats.Timestamp.UnixNano()
+
+	// zero or negative interval
+	if nanoSeconds <= 0 {
+		return 0
+	}
```

Comment:
We also need to check that the stats didn't decrease: https://github.com/containerd/containerd/pull/10652  and that test case in